Sources/Fluent/ModelCredentialsAuthenticatable.swift

```diff
@@ -0,0 +1,54 @@
+import Vapor
+
+public protocol ModelCredentialsAuthenticatable: Model, Authenticatable {
+    static var usernameKey: KeyPath<Self, Field<String>> { get }
+    static var passwordHashKey: KeyPath<Self, Field<String>> { get }
+    func verify(password: String) throws -> Bool
+}
+
+extension ModelCredentialsAuthenticatable {
+    public static func credentialsAuthenticator(
+        database: DatabaseID? = nil
+    ) -> Authenticator {
+        ModelCredentialsAuthenticator<Self>(database: database)
+    }
+
+    var _$username: Field<String> {
+        self[keyPath: Self.usernameKey]
+    }
+
+    var _$passwordHash: Field<String> {
+        self[keyPath: Self.passwordHashKey]
+    }
+}
+
+public struct ModelCredentials: Content {
+    public let username: String
+    public let password: String
+
+    public init(username: String, password: String) {
+        self.username = username
+        self.password = password
+    }
+}
+
+private struct ModelCredentialsAuthenticator<User>: CredentialsAuthenticator
+    where User: ModelCredentialsAuthenticatable
+{
+    typealias Credentials = ModelCredentials
+
+    public let database: DatabaseID?
+
+    func authenticate(credentials: ModelCredentials, for request: Request) -> EventLoopFuture<Void> {
+        User.query(on: request.db(self.database)).filter(\._$username == credentials.username).first().flatMapThrowing { foundUser in
```

Comment:
Maybe, since this is a default implementation that a lot of people will end up using, we should do something like this:
```swift
    .all()
    .flatMapThrowing {
#if DEBUG
        assert($0.count <= 1 , "Usernames must be unique in your database")
#else
        guard $0.count <= 1 else throw Abort(.conflict, reason: "Found more than one matching user")
#endif
        return $0.first
    }
```
I'd hate for it to turn into another potential footgun for the unsuspecting 😕. Granted, this would only catch the case where someone actually _did_ end up with multiple of the same username, but at least it's something.

Reply:
Given this affects all of the authenticators, we should probably do that as a separate PR